File: leetcode/438. Find All Anagrams in a String/438.py
// https://helloacm.com/algorithm-of-two-pointer-sliding-windows-to-find-all-anagrams-in-a-string/
// https://leetcode.com/problems/find-all-anagrams-in-a-string/

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p) > len(s):
            return []
        freq1 = collections.Counter(s[:len(p)])
        freq2 = collections.Counter(p)
        r = []
        if freq1 == freq2:
            r.append(0)
        i = len(p)
        while i < len(s):  
            if s[i] in freq1:
                freq1[s[i]] += 1
            else:
                freq1[s[i]] = 1
            freq1[s[i - len(p)]] -= 1
            if freq1[s[i - len(p)]] == 0:
                del freq1[s[i - len(p)]]
            if freq1 == freq2:
                r.append(i - len(p) + 1)
            i += 1
        return r            
        

File: leetcode/438. Find All Anagrams in a String/438.cpp
// https://helloacm.com/algorithm-of-two-pointer-sliding-windows-to-find-all-anagrams-in-a-string/
// https://leetcode.com/problems/find-all-anagrams-in-a-string/

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if (p.size() > s.size()) return {};
        int freq1[26] = { 0 };
        int freq2[26] = { 0 };
        for (int i = 0; i < p.size(); ++ i) {
            char n = s[i];
            freq1[n - 'a'] ++;
        }
        for (const auto &n: p) {
            freq2[n - 'a'] ++;
        }
        vector<int> r;                
        if (equal(freq1, freq2)) { // anagrams at index 0
            r.push_back(0);
        }
        for (int i = p.size(); i < s.size(); ++ i) {
            freq1[s[i] - 'a'] ++; // new character
            freq1[s[i - p.size()] - 'a'] --; // remove the first character
            if (equal(freq1, freq2)) {
                r.push_back(i - p.size() + 1);
            }
        }
        return r;
    }
 
    // O(1) to check if two frequency table matches.    
    bool equal(int *freq1, int *freq2) {
        for (int i = 0; i < 26; ++ i) {
            if (freq1[i] != freq2[i]) return false;
        }
        return true;
    }
};
